File: packages/monitorme-server-agent-py/monitorme_server_agent_py/tracing.py
from __future__ import annotations

from opentelemetry import trace
from opentelemetry.exporter.zipkin.json import ZipkinExporter
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor
from opentelemetry.sdk.resources import SERVICE_NAME, Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

from .config import AgentConfig
from .middleware import MonitorMeASGIBaggageMiddleware


def _spans_endpoint(endpoint: str) -> str:
    return f"{endpoint.rstrip('/')}/spans"


def setup_tracing(config: AgentConfig) -> TracerProvider:
    headers = {}
    if config.auth_token:
        headers["Authorization"] = f"Bearer {config.auth_token}"

    resource = Resource.create({SERVICE_NAME: config.service_name})
    tracer_provider = TracerProvider(resource=resource)
    tracer_provider.add_span_processor(
        BatchSpanProcessor(
            ZipkinExporter(
                endpoint=_spans_endpoint(config.endpoint),
                headers=headers,
            )
        )
    )
    trace.set_tracer_provider(tracer_provider)
    RequestsInstrumentor().instrument()
    return tracer_provider


def setup_fastapi(app, config: AgentConfig) -> TracerProvider:
    tracer_provider = setup_tracing(config)
    app.add_middleware(MonitorMeASGIBaggageMiddleware)
    FastAPIInstrumentor.instrument_app(app, tracer_provider=tracer_provider)
    return tracer_provider

File: packages/monitorme-server-agent-js/tracing.js
"use strict";

const { NodeTracerProvider } = require("@opentelemetry/node");
const { registerInstrumentations } = require("@opentelemetry/instrumentation");
const { ZipkinExporter } = require("@opentelemetry/exporter-zipkin");
const { Resource } = require("@opentelemetry/resources");
const { ResourceAttributes } = require("@opentelemetry/semantic-conventions");
const config = require("./config.json");

// Import the utilities from utils.js
const {
  EnhancedSpanProcessor,
  createAndRegisterInstrumentations,
} = require("./utils");

// Create a ZipkinExporter to send trace data to Zipkin
const traceExporter = new ZipkinExporter({
  url: `${config.endpoint}/spans`, // Define the endpoint where traces will be sent
  headers: {
    Authorization: `Bearer ${config.authTokeh}`,
  },
});

// Set up the tracer provider with resources, such as service name
const tracerProvider = new NodeTracerProvider({
  resource: new Resource({
    [ResourceAttributes.SERVICE_NAME]: config.serviceName, // Set service name from configuration
  }),
});

// Add the custom EnhancedSpanProcessor to handle spans before exporting them
tracerProvider.addSpanProcessor(
  new EnhancedSpanProcessor(traceExporter, {
    maxQueueSize: 10000, // Maximum number of spans that can be queued
    maxExportBatchSize: 5120, // Maximum number of spans that can be exported in one batch
    scheduledDelayMillis: 5000, // Time interval between export attempts
    exportTimeoutMillis: 30000, // Maximum time allowed for exporting a batch of spans
  }),
);

// Register the provider to make it active for instrumentation
tracerProvider.register();

// Register the instrumentations from the utils function
registerInstrumentations({
  instrumentations: createAndRegisterInstrumentations(config.dbOptions),
});
